sky/backends/backend_utils.py

```diff
@@ -990,6 +1000,119 @@ def get_ready_nodes_counts(pattern, output):
     return ready_head, ready_workers
 
 
+@timeline.event
+def _deterministic_yaml_hash(yaml_path: str) -> str:
+    """Hashes the cluster yaml and contents of file mounts. Two invocations of
+    this function should return the same string if and only if the contents of
+    the yaml are the same and the file contents of all the file_mounts specified
+    in the yaml are the same.
+
+    Limitations:
+    - This function can be expensive if the file mounts are large. (E.g. a few
+      seconds for ~1GB.)
+    - Symbolic links are not explicitly handled. Some symbolic link changes may
+      not be detected.
+
+    Implementation: We create a byte sequence that captures the state of the
+    yaml file and all the files in the file mounts, then hash the byte sequence.
+
+    The format of the byte sequence is:
+    32 bytes - sha256 hash of the yaml file
+    for each file mount:
+      file mount remote destination (UTF-8), \0
+      if the file mount source is a file:
+        'file' encoded to UTF-8
+        32 byte sha256 hash of the file contents
+      if the file mount source is a directory:
+        'dir' encoded to UTF-8
+        for each directory and subdirectory withinin the file mount (starting from
+            the root and descending recursively):
+          name of the directory (UTF-8), \0
+          name of each subdirectory within the directory (UTF-8) terminated by \0
+          \0
+          for each file in the directory:
+            name of the file (UTF-8), \0
+            32 bytes - sha256 hash of the file contents
+          \0
+      if the file mount source is something else or does not exist, nothing
+      \0\0
+
+    Rather than constructing the whole byte sequence, which may be quite large,
+    we construct it incrementally by using hash.update() to add new bytes.
+    """
+
+    # In python 3.11, hashlib.file_digest is available, but for <3.11 we have to
+    # do it manually.
+    # This implementation is simplified from the implementation in CPython.
+    # Beware of f.read() as some files may be larger than memory.
+    def _hash_file(path: str) -> bytes:
+        with open(path, 'rb') as f:
+            file_hash = hashlib.sha256()
+            buf = bytearray(2**18)
+            view = memoryview(buf)
+            while True:
+                size = f.readinto(buf)
+                if size == 0:
+                    # EOF
+                    break
+                file_hash.update(view[:size])
+            return file_hash.digest()
+
+    config_hash = hashlib.sha256()
+
+    config_hash.update(_hash_file(yaml_path))
+
+    yaml_config = common_utils.read_yaml(yaml_path)
+    file_mounts = yaml_config.get('file_mounts', {})
+    # Remove the file mounts added by the newline.
+    if '' in file_mounts:
+        assert file_mounts[''] == '', file_mounts['']
+        file_mounts.pop('')
+
+    for dst, src in sorted(file_mounts.items()):
+        expanded_src = os.path.expanduser(src)
```

Comment:
I don't think _optimize_file_mounts or the actual ray rsync will resolve symlinks, so I opted not to here.

Reply:
I'll add a more comprehensive comment about symlink behavior:
- we do not resolve the path of file mounts that are symlinks (the hashed file path is the symlink path, not the resolved path)
- if the file mount is a symlink to a file, we will hash the file contents
- if the file mount is a symlink to a dir, we will hash the contents of dir as normal
  - if the passed dir to os.walk is a symlinked, it will resolve that link
- if a directory includes a symlink to a dir, we will hash the unresolved symlink path but will not recurse into the dir
  - os.walk will not recurse into symlinked directories but will include them in the directories list
- if a directory includes a symlink to a file, we will hash the file contents